### PR TITLE
Bringing SQLCipher support to OS X, Windows, and Linux

### PR DESCRIPTION
**Pull request by borrrden**
Hello!  I've been hard at work for the past week or two extending this library to bring SQLCipher support to all the platforms that Couchbase Lite targets (iOS, Android, OS X, Linux, Windows).  I thought this might be useful to you, so I will summarize what I've done:

1) Getting SQLCipher native image built.  This was tricky, especially on Windows.  However, in the end, I realized that the configure script simply generates a sqlite.c amalgamation including the crypto stuff so I just committed the amalgamation to skip that cumbersome step.  It is based on SQLite 3.8.10.2.  This way, in order to upgrade in the future you just need to make a new amalgamation.  I then set up a script to build for each platform (two for Android, one for *nix and one for Windows).  I did away with OpenSSL, and instead used CommonCrypto for OS X and iOS (since it is a system lib).  I used libtomcrypt for the others since building OpenSSL is harder than it should be.  This process is annoying, so I also included prebuilt images for each platform / architecture (The exception being that Linux only includes a 64-bit image since I cannot figure out how to do the 32-bit/64-bit trick that Windows does in any sane way).  However, Linux folks are usually comfortable modifying so I think they will not have a problem changing the script to make a 32-bit build if they need it.  

2) Some changes to the pinvoke.  I needed the x86/x64 trick for the regular sqlite3.dll on Windows to also work for packaged_sqlcipher.dll.  Also, iOS no longer needs to link to libcrypto.a.  All of the other stuff has to do with .NET 3.5 compatibility and you are free to take it or leave it.  

For my fork I generated the project files and committed the ones I am using, but I assume you will want to build the new platform packaged sqlcipher targets into your gen_build.cs.  I'll leave that part up to you.

### CODE REVIEW

**Comment by ericsink** [comment]
Neato.

I'll look more closely at this.

E

File: src/cs/sqlite3_pinvoke.cs [hunk 1]
```diff
@@ -53,21 +53,21 @@
...
+        LinkTarget = ObjCRuntime.LinkTarget.Simulator | ObjCRuntime.LinkTarget.Simulator64 | ObjCRuntime.LinkTarget.ArmV7 | ObjCRuntime.LinkTarget.Arm64,
         ForceLoad=true,
         LinkerFlags="",
-        Frameworks=""
+        Frameworks="Security"
         )
         ]
 
-[assembly: ObjCRuntime.LinkWith(
+/*[assembly: ObjCRuntime.LinkWith(
```
Comment (nberardi) [inline src/cs/sqlite3_pinvoke.cs]: I would remove the commented out code.

File: src/cs/sqlite3_pinvoke.cs [hunk 2]
```diff
@@ -1218,15 +1220,15 @@ private static class NativeMethods
 
         private static bool TryLoadFromDirectory(string dllName, string baseDirectory)
         {
-            System.Diagnostics.Debug.Assert(!string.IsNullOrWhiteSpace(dllName), "dllName is null or empty.");
```
Comment (nberardi) [inline src/cs/sqlite3_pinvoke.cs]: Why remove these? a white space is not a valid `dllName` or `baseDirectory`

File: src/cs/sqlite3_pinvoke.cs [hunk 3]
```diff
@@ -1218,15 +1220,15 @@ private static class NativeMethods
...
+            System.Diagnostics.Debug.Assert(!string.IsNullOrEmpty(dllName), "dllName is null or empty.");
+            System.Diagnostics.Debug.Assert(!string.IsNullOrEmpty(baseDirectory), "baseDirectory is null or empty.");
             System.Diagnostics.Debug.Assert(System.IO.Path.IsPathRooted(baseDirectory), "baseDirectory is not rooted.");
 
             var architecture = IntPtr.Size == 4
                 ? "x86"
                 : "x64";
 
-            var dllPath = System.IO.Path.Combine(baseDirectory, architecture, dllName);
+            var dllPath = System.IO.Path.Combine(System.IO.Path.Combine(baseDirectory, architecture), dllName);
```
Comment (nberardi) [inline src/cs/sqlite3_pinvoke.cs]: A `Path.Combine` inside another `Path.Combine` offers no value.

File: src/cs/sqlite3_pinvoke.cs (hunk 2 above)
Comment (borrrden) [inline src/cs/sqlite3_pinvoke.cs]: That API is not available in .NET 3.5 (the platform I am targeting). So you can take this change or leave it :)

File: src/cs/sqlite3_pinvoke.cs (hunk 3 above)
Comment (borrrden) [inline src/cs/sqlite3_pinvoke.cs]: The Params version is not available in .NET 3.5.

**Comment by ericsink** [comment]
Sorry I haven't dug into this deeply yet.  Crazy week.

**Comment by ericsink** [comment]
Update:  The major issue here is the need to update gen_build.cs to generate the necessary project files.  I've got similar stuff to be done for UWP support as well, so I am trying to do both while I am in that area of the code.

**Comment by borrrden** [comment]
I'm ok with this not being here yet.  In fact I may make some changes to it because we don't use it like this anymore.  Although to use our current way you would need to take a dependency on our repo (we have a dedicated repo for building and housing libcrypto and native sqlcipher).  I guess the question is would you be open to that?

**Comment by ericsink** [comment]
I like to avoid dependencies, but in this case, yes, I would be open to taking a dependency on your work for the crypto and sqlcipher stuff.

Now you've got me kind of curious about how your use of these items has changed since this PR.

**Comment by borrrden** [comment]
The big thing is that I gave you a native build system with LibTomCrypt, but now we use OpenSSL as you did before (except on Apple products where we use CommonCrypto).  This is so that we can use a FIPS compliant crypto module if we ever need to (OpenSSL can be set that way and CommonCrypto always is).

[This](https://github.com/couchbaselabs/couchbase-lite-libsqlcipher) is the repo for our native SQLCipher.  It takes a dep on the official sqlcipher repo, as well as another repo we have for building libcrypto (We use that for our other native storage engine as well, which is why it is separate).  Those two links are in the vendor folder.

**Comment by ericsink** [comment]
This stuff got integrated into master in a different way.

**Pull request closed without merging**